getMapping fails future when index not found

Execution of
```scala
client.execute {
      getMapping(indexName / indexType)
 }
```
**in case when there is no such index** returns Failed Future with:
```
Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize instance of `scala.collection.immutable.Map` out of START_ARRAY token
 at [Source: (String)"{"error":{"root_cause":[{"type":"index_not_found_exception","reason":"no such index","resource.type":"index_or_alias","resource.id":"247acbc1-ff7a-4f4c-bcb2-23ec1e56ccc2","index_uuid":"_na_","index":"247acbc1-ff7a-4f4c-bcb2-23ec1e56ccc2"}],"type":"index_not_found_exception","reason":"no such index","resource.type":"index_or_alias","resource.id":"247acbc1-ff7a-4f4c-bcb2-23ec1e56ccc2","index_uuid":"_na_","index":"247acbc1-ff7a-4f4c-bcb2-23ec1e56ccc2"},"status":404}"; line: 1, column: 24] (through reference chain: com.fasterxml.jackson.module.scala.deser.MapBuilderWrapper["error"]->com.fasterxml.jackson.module.scala.deser.MapBuilderWrapper["root_cause"])
	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:63)
	at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1342)
	at com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1138)
	at com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1092)
	at com.fasterxml.jackson.databind.deser.std.StdDeserializer._deserializeFromEmpty(StdDeserializer.java:599)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:360)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:54)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:38)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBindStringKeyMap(MapDeserializer.java:527)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:364)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:54)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:38)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBindStringKeyMap(MapDeserializer.java:527)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:364)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:54)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:38)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4001)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3030)
	at com.fasterxml.jackson.module.scala.experimental.ScalaObjectMapper$class.readValue(ScalaObjectMapper.scala:190)
	at com.sksamuel.elastic4s.json.JacksonSupport$$anon$1.readValue(JacksonSupport.scala:11)
	at com.sksamuel.elastic4s.http.ResponseHandler$.fromEntity(ResponseHandler.scala:35)
	at com.sksamuel.elastic4s.http.ResponseHandler$.fromResponse(ResponseHandler.scala:30)
	at com.sksamuel.elastic4s.http.index.mappings.MappingExecutables$GetMappingHttpExecutable$$anon$1.handle(MappingExecutables.scala:19)
	at com.sksamuel.elastic4s.http.index.mappings.MappingExecutables$GetMappingHttpExecutable$$anon$1.handle(MappingExecutables.scala:17)
	at com.sksamuel.elastic4s.http.HttpClient$$anonfun$execute$1.apply(HttpClient.scala:69)
	at com.sksamuel.elastic4s.http.HttpClient$$anonfun$execute$1.apply(HttpClient.scala:68)
	at scala.util.Success$$anonfun$map$1.apply(Try.scala:237)
	at scala.util.Try$.apply(Try.scala:192)
	at scala.util.Success.map(Try.scala:237)
	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:237)
	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:237)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:36)
	at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:121)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

I would expect `RequestFailure` in the Either of the response.

--- a/elastic4s-http/src/main/scala/com/sksamuel/elastic4s/http/index/mappings/MappingExecutables.scala
+++ b/elastic4s-http/src/main/scala/com/sksamuel/elastic4s/http/index/mappings/MappingExecutables.scala
@@ -1,7 +1,7 @@
 package com.sksamuel.elastic4s.http.index.mappings
 
 import com.sksamuel.elastic4s.IndexesAndTypes
-import com.sksamuel.elastic4s.http.{HttpEntity, HttpExecutable, HttpRequestClient, HttpResponse, ResponseHandler}
+import com.sksamuel.elastic4s.http._
 import com.sksamuel.elastic4s.indexes.PutMappingBuilderFn
 import com.sksamuel.elastic4s.mappings.{GetMappingDefinition, PutMappingDefinition}
 import org.apache.http.entity.ContentType
@@ -15,18 +15,21 @@ trait MappingExecutables {
   implicit object GetMappingHttpExecutable extends HttpExecutable[GetMappingDefinition, Seq[IndexMappings]] {
 
     override def responseHandler: ResponseHandler[Seq[IndexMappings]] = new ResponseHandler[Seq[IndexMappings]] {
-      override def handle(response: HttpResponse) = {
-        val raw = ResponseHandler.fromResponse[Map[String, Map[String, Map[String, Map[String, Any]]]]](response)
-        val raw2 = raw.map {
-          case (index, types) =>
-            val mappings = types("mappings").map {
-              case (tpe, properties) =>
-                tpe -> properties.get("properties").map(_.asInstanceOf[Map[String, Any]]).getOrElse(Map.empty)
-            }
-            IndexMappings(index, mappings)
-        }.toSeq
-        Right(raw2)
-      }
+      override def handle(response: HttpResponse): Either[ElasticError, Seq[IndexMappings]] =
+        response.statusCode match {
+          case 201 | 200 =>
+            val raw = ResponseHandler.fromResponse[Map[String, Map[String, Map[String, Map[String, Any]]]]](response)
+            val raw2 = raw.map {
+              case (index, types) =>
+                val mappings = types("mappings").map {
+                  case (tpe, properties) =>
+                    tpe -> properties.get("properties").map(_.asInstanceOf[Map[String, Any]]).getOrElse(Map.empty)
+                }
+                IndexMappings(index, mappings)
+            }.toSeq
+            Right(raw2)
+          case _ => Left(ElasticError.parse(response))
+        }
     }
 
     override def execute(client: HttpRequestClient, request: GetMappingDefinition): Future[HttpResponse] = {

--- a/elastic4s-tests/src/test/scala/com/sksamuel/elastic4s/mappings/MappingHttpTest.scala
+++ b/elastic4s-tests/src/test/scala/com/sksamuel/elastic4s/mappings/MappingHttpTest.scala
@@ -67,5 +67,14 @@ class MappingHttpTest extends WordSpec with DockerTests with Matchers {
 
       properties shouldBe Map.empty
     }
+
+    "handle error response" in {
+      val result = http.execute {
+        getMapping("nonexistent" / "mapping2")
+      }.await
+
+      result.isLeft shouldBe true
+      result.left.get.error.`type` shouldBe "index_not_found_exception"
+    }
   }
 }

